algo/src/main/java/org/neo4j/graphalgo/impl/HugeLabelPropagation.java

```diff
@@ -0,0 +1,240 @@
+                    existingLabels,
+                    nodes,
+                    randomizeOrder
+            );
+        }
+    }
+
+    private static final class ComputeStep extends Computation implements HugeRelationshipConsumer {
+
+        private final HugeRelationshipIterator graph;
+        private final HugeRelationshipWeights relationshipWeights;
+        private final HugeWeightMapping nodeWeights;
+        private final ProgressLogger progressLogger;
+        private final Direction direction;
+        private final long maxNode;
+        private final HugeLongArray existingLabels;
+        private final PrimitiveLongIterable nodes;
+        private final LongDoubleHashMap votes;
+
+        private ComputeStep(
+                HugeRelationshipIterator graph,
+                HugeRelationshipWeights relationshipWeights,
+                HugeWeightMapping nodeWeights,
+                ProgressLogger progressLogger,
+                Direction direction,
+                final long maxNode,
+                HugeLongArray existingLabels,
+                PrimitiveLongIterable nodes,
+                boolean randomizeOrder) {
+            this.graph = graph;
+            this.relationshipWeights = relationshipWeights;
+            this.nodeWeights = nodeWeights;
+            this.progressLogger = progressLogger;
+            this.direction = direction;
+            this.maxNode = maxNode;
+            this.existingLabels = existingLabels;
+            this.nodes = RandomlySwitchingLongIterable.of(randomizeOrder, nodes);
+            this.votes = new LongDoubleScatterMap();
+        }
+
+        @Override
+        boolean computeAll() {
+            PrimitiveLongIterator iterator = nodes.iterator();
+            boolean didChange = false;
+            while (iterator.hasNext()) {
+                didChange = compute(iterator.next(), didChange);
+            }
+            return didChange;
+        }
+
+        private boolean compute(long nodeId, boolean didChange) {
+            votes.clear();
+            long partition = existingLabels.get(nodeId);
+            long previous = partition;
+            graph.forEachRelationship(nodeId, direction, this);
+            double weight = Double.NEGATIVE_INFINITY;
+            for (LongDoubleCursor vote : votes) {
+                if (weight < vote.value) {
+                    weight = vote.value;
+                    partition = vote.key;
+                }
+            }
+            progressLogger.logProgress(nodeId, maxNode);
+            if (partition != previous) {
+                existingLabels.set(nodeId, partition);
+                return true;
+            }
+            return didChange;
+        }
+
+        @Override
+        public boolean accept(final long sourceNodeId, final long targetNodeId) {
+            long partition = existingLabels.get(targetNodeId);
+            double relationshipWeight = relationshipWeights.weightOf(sourceNodeId, targetNodeId);
```

Comment:
these two are probably the most expensive operations.

in the other code for pagerank we had a cache for rel-weights and node-weights and used the variant of the consumer that got the rel-weight passed in.

Reply:
`existingLabels.get` is basically an array lookup, two at the most; `relationshipWeights.weightOf` is one or two array lookups followed by a hash-map lookup; `nodeWeights.nodeWeight` is an array lookup followed by a hash-map lookup. With caching similar to what Page Rank is doing (can't do the exact same thing, as we're in huge land with long keys and direct array caches don't work anymore, requiring something like the HugeLongArray, that is backing the existingLabels, which is the closest I got to plain primitive arrays; see LongArrayBenchmark) we could get rid of the map lookups but might have to inflate the memory usage since we would have materialize all default values. On the other hand – If we have >50% of rels with a non-default weight, we actually save storage compared to the map. I would rather suggest to have a better weight mapping implementation, that would switch the internal storage from a map to an array if we have enough values and that would then be close enough to the cache so that we don't need it anymore.

Reply:
Ok, makes sense, we can track this as a separate issue. Thanks!